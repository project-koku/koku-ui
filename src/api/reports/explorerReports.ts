import type { ReportItem } from './report';

export interface ExplorerReportItem extends ReportItem {
  account?: string;
  account_alias?: string;
  cluster?: string;
  gcp_project?: string;
  node?: string;
  org_unit_id?: string;
  payer_tenant_id?: string;
  product_service?: string;
  project?: string;
  region?: string;
  resource_location?: string;
  service?: string;
  service_name?: string;
  subscription_guid?: string;
}
