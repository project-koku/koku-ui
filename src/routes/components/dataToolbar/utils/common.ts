import type { ToolbarLabelGroup } from '@patternfly/react-core';
import type { Query } from 'api/queries/query';
import { intl } from 'components/i18n';
import messages from 'locales/messages';
import { cloneDeep } from 'lodash';
import type { Filter } from 'routes/utils/filter';
import { awsCategoryKey, awsCategoryPrefix, orgUnitIdKey, tagKey, tagPrefix } from 'utils/props';

export interface Filters {
  [key: string]: Filter[] | { [key: string]: Filter[] };
}

export interface ToolbarChipGroupExt extends ToolbarLabelGroup {
  ariaLabelKey?: string;
  placeholderKey?: string;
  resourceKey?: string;
  selectClassName?: string; // A selector from routes/components/dataToolbar/dataToolbar.scss
  selectOptions?: ToolbarLabelGroup[];
}

export const defaultFilters = {
  [awsCategoryKey]: {},
  tag: {},
};

// Remove trailing commas -- see https://issues.redhat.com/browse/COST-3641
export const cleanInput = (value: string) => {
  return value.replace(/,*$/g, '').replace(/&*$/g, '');
};

export const getActiveFilters = query => {
  const filters = cloneDeep(defaultFilters);

  const parseFilters = (key, values, isExcludes = false) => {
    if (key.indexOf(tagPrefix) !== -1) {
      if (filters.tag[key.substring(tagPrefix.length)]) {
        filters.tag[key.substring(tagPrefix.length)] = [
          ...filters.tag[key.substring(tagPrefix.length)],
          ...getFilters(key, values, isExcludes),
        ];
      } else {
        filters.tag[key.substring(tagPrefix.length)] = getFilters(key, values, isExcludes);
      }
    } else if (key.indexOf(awsCategoryPrefix) !== -1) {
      if (filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)]) {
        filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)] = [
          ...filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)],
          ...getFilters(key, values, isExcludes),
        ];
      } else {
        filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)] = getFilters(key, values, isExcludes);
      }
    } else if (filters[key]) {
      filters[key] = [...filters[key], ...getFilters(key, values, isExcludes)];
    } else {
      filters[key] = getFilters(key, values, isExcludes);
    }
  };

  if (query?.filter_by) {
    Object.keys(query.filter_by).forEach(key => {
      const values = Array.isArray(query.filter_by[key]) ? [...query.filter_by[key]] : [query.filter_by[key]];
      parseFilters(key, values);
    });
  }
  if (query?.exclude) {
    Object.keys(query.exclude).forEach(key => {
      const values = Array.isArray(query.exclude[key]) ? [...query.exclude[key]] : [query.exclude[key]];
      parseFilters(key, values, true);
    });
  }
  return filters;
};

export const getChips = (filters: Filter[]): string[] => {
  const chips = [];
  if (filters instanceof Array) {
    filters.forEach(item => {
      const label = item.toString ? item.toString() : undefined;

      chips.push({
        key: item.value,
        node: label
          ? label
          : item.isExcludes
            ? intl.formatMessage(messages.excludeLabel, { value: item.value })
            : item.value,
      });
    });
  }
  return chips;
};

export const getDefaultCategory = (categoryOptions: ToolbarLabelGroup[], groupBy: string, query: Query) => {
  if (!categoryOptions) {
    return 'name';
  }
  if (query && query.group_by && query.group_by[orgUnitIdKey]) {
    return orgUnitIdKey;
  }
  for (const option of categoryOptions) {
    if (groupBy === option.key || (groupBy && groupBy.indexOf(tagPrefix) !== -1 && option.key === tagKey)) {
      return option.key;
    }
  }
  return categoryOptions[0].key;
};

export const getFilter = (
  filterType: string,
  filterValue: string,
  isExcludes = false,
  toString = undefined
): Filter => {
  return { type: filterType, value: filterValue, isExcludes, toString };
};

export const getFilters = (filterType: string, filterValues: string[], isExcludes = false): Filter[] => {
  return filterValues.map(value => getFilter(filterType, value, isExcludes));
};

export const hasFilters = (filters: Filters) => {
  if (filters) {
    for (const filterKey of Object.keys(filters)) {
      if (filterKey === awsCategoryKey || filterKey === tagKey) {
        for (const key of Object.keys(filters[filterKey])) {
          if (filters[filterKey][key]) {
            return true;
          }
        }
      } else {
        return true;
      }
    }
  }
  return false;
};

export const onDelete = (type: any, chip: any, currentFilters) => {
  let filter = null; // Clear all
  let filters = cloneDeep(defaultFilters);

  // Todo: workaround for https://github.com/patternfly/patternfly-react/issues/3552
  // This prevents us from using a localized string, if necessary
  let _type = type && type.key ? type.key : type;
  if (_type && _type.indexOf(tagPrefix) !== -1) {
    _type = _type.slice(tagPrefix.length);
  } else if (_type && _type.indexOf(awsCategoryPrefix) !== -1) {
    _type = _type.slice(awsCategoryPrefix.length);
  }

  if (_type) {
    const excludePrefix = intl.formatMessage(messages.excludeLabel, { value: '' });
    let id = chip && chip.key ? chip.key : chip;
    if (id && id.indexOf(excludePrefix) !== -1) {
      const isExcludes = id ? id.indexOf(excludePrefix) !== -1 : false;
      id = isExcludes ? id.slice(excludePrefix.length) : id;
    }
    if (currentFilters.tag[_type]) {
      filter = currentFilters.tag[_type].find(item => item.value === id);
    } else if (currentFilters[awsCategoryKey][_type]) {
      filter = currentFilters[awsCategoryKey][_type].find(item => item.value === id);
    } else if (currentFilters[_type]) {
      filter = (currentFilters[_type] as Filter[]).find(item => item.value === id);
    }

    const newFilters: any = cloneDeep(currentFilters);
    if (newFilters.tag[_type]) {
      // Todo: use ID
      newFilters.tag[_type] = newFilters.tag[_type].filter(item => item.value !== id);
    } else if (newFilters[awsCategoryKey][_type]) {
      newFilters[awsCategoryKey][_type] = newFilters[awsCategoryKey][_type].filter(item => item.value !== id);
    } else if (newFilters[_type]) {
      newFilters[_type] = newFilters[_type].filter(item => item.value !== id);
    }
    filters = newFilters;
  }
  return {
    filter,
    filters,
  };
};
