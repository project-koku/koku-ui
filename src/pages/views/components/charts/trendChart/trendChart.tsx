import 'pages/views/components/charts/common/charts-common.scss';

import {
  Chart,
  ChartArea,
  ChartAxis,
  ChartLegend,
  ChartLegendTooltip,
  createContainer,
  getInteractiveLegendEvents,
} from '@patternfly/react-charts';
import { Title } from '@patternfly/react-core';
import { getDate } from 'date-fns';
import messages from 'locales/messages';
import { default as ChartTheme } from 'pages/views/components/charts/chartTheme';
import { getCostRangeString, getDateRange } from 'pages/views/components/charts/common/chartDatumUtils';
import {
  ChartSeries,
  getChartNames,
  getDomain,
  getLegendData,
  getResizeObserver,
  getTooltipLabel,
  initHiddenSeries,
  isDataAvailable,
  isSeriesHidden,
} from 'pages/views/components/charts/common/chartUtils';
import React from 'react';
import { injectIntl, WrappedComponentProps } from 'react-intl';
import { FormatOptions, Formatter } from 'utils/format';
import { noop } from 'utils/noop';

import { chartStyles } from './trendChart.styles';

interface TrendChartOwnProps {
  adjustContainerHeight?: boolean;
  containerHeight?: number;
  currentData: any;
  forecastData?: any;
  forecastConeData?: any;
  formatOptions?: FormatOptions;
  formatter: Formatter;
  height?: number;
  legendItemsPerRow?: number;
  previousData?: any;
  padding?: any;
  showForecast?: boolean; // Show forecast legend regardless if data is available
  showInfrastructureLabel?: boolean; // Show supplementary cost labels
  showSupplementaryLabel?: boolean; // Show supplementary cost labels
  showUsageLegendLabel?: boolean; // The cost legend label is shown by default
  title?: string;
  units?: string;
}

interface State {
  cursorVoronoiContainer?: any;
  hiddenSeries: Set<number>;
  series?: ChartSeries[];
  width: number;
}

type TrendChartProps = TrendChartOwnProps & WrappedComponentProps;

class TrendChartBase extends React.Component<TrendChartProps, State> {
  private containerRef = React.createRef<HTMLDivElement>();
  private observer: any = noop;

  public state: State = {
    hiddenSeries: new Set(),
    width: 0,
  };

  public componentDidMount() {
    this.initDatum();
    this.observer = getResizeObserver(this.containerRef.current, this.handleResize);
  }

  public componentDidUpdate(prevProps: TrendChartProps) {
    if (
      prevProps.currentData !== this.props.currentData ||
      prevProps.forecastData !== this.props.forecastData ||
      prevProps.forecastConeData !== this.props.forecastConeData ||
      prevProps.previousData !== this.props.previousData
    ) {
      this.initDatum();
    }
  }

  public componentWillUnmount() {
    if (this.observer) {
      this.observer();
    }
  }

  private initDatum = () => {
    const {
      currentData,
      forecastData,
      forecastConeData,
      previousData,
      showForecast,
      showInfrastructureLabel = false,
      showSupplementaryLabel = false,
      showUsageLegendLabel = false,
    } = this.props;

    const key = showUsageLegendLabel
      ? messages.ChartUsageLegendLabel
      : showSupplementaryLabel
      ? messages.ChartCostSupplementaryLegendLabel
      : showInfrastructureLabel
      ? messages.ChartCostInfrastructureLegendLabel
      : messages.ChartCostLegendLabel;

    const tooltipKey = showUsageLegendLabel
      ? messages.ChartUsageLegendTooltip
      : showSupplementaryLabel
      ? messages.ChartCostSupplementaryLegendTooltip
      : showInfrastructureLabel
      ? messages.ChartCostInfrastructureLegendTooltip
      : messages.ChartCostLegendTooltip;

    const noDataKey = showUsageLegendLabel
      ? messages.ChartUsageLegendNoDataLabel
      : showSupplementaryLabel
      ? messages.ChartCostSupplementaryLegendNoDataLabel
      : showInfrastructureLabel
      ? messages.ChartCostInfrastructureLegendNoDataLabel
      : messages.ChartCostLegendNoDataLabel;

    // Show all legends, regardless of length -- https://github.com/project-koku/koku-ui/issues/248

    const series: ChartSeries[] = [
      {
        childName: 'previousCost',
        data: previousData,
        legendItem: {
          name: getCostRangeString(previousData, key, true, true, 1, noDataKey),
          symbol: {
            fill: chartStyles.previousColorScale[0],
            type: 'minus',
          },
          tooltip: getCostRangeString(previousData, tooltipKey, false, false, 1),
        },
        style: {
          data: {
            ...chartStyles.previousMonthData,
            stroke: chartStyles.previousColorScale[0],
          },
        },
      },
      {
        childName: 'currentCost',
        data: currentData,
        legendItem: {
          name: getCostRangeString(currentData, key, true, false, 0, noDataKey),
          symbol: {
            fill: chartStyles.currentColorScale[0],
            type: 'minus',
          },
          tooltip: getCostRangeString(currentData, tooltipKey, false, false),
        },
        style: {
          data: {
            ...chartStyles.currentMonthData,
            stroke: chartStyles.currentColorScale[0],
          },
        },
      },
    ];

    if (showForecast) {
      series.push({
        childName: 'forecast',
        data: forecastData,
        legendItem: {
          name: getCostRangeString(
            forecastData,
            messages.ChartCostForecastLegendLabel,
            false,
            false,
            0,
            messages.ChartCostForecastLegendNoDataLabel
          ),
          symbol: {
            fill: chartStyles.forecastDataColorScale[0],
            type: 'minus',
          },
          tooltip: getCostRangeString(forecastData, messages.ChartCostForecastLegendTooltip, false, false),
        },
        style: {
          data: {
            ...chartStyles.forecastData,
            stroke: chartStyles.forecastDataColorScale[0],
          },
        },
      });
      series.push({
        childName: 'forecastCone',
        data: forecastConeData,
        legendItem: {
          name: getCostRangeString(
            forecastConeData,
            messages.ChartCostForecastConeLegendLabel,
            false,
            false,
            0,
            messages.ChartCostForecastConeLegendNoDataLabel
          ),
          symbol: {
            fill: chartStyles.forecastConeDataColorScale[0],
            type: 'triangleLeft',
          },
          tooltip: getCostRangeString(forecastConeData, messages.ChartCostForecastConeLegendTooltip, false, false),
        },
        style: {
          data: {
            ...chartStyles.forecastConeData,
            stroke: chartStyles.forecastConeDataColorScale[0],
          },
        },
      });
    }
    const cursorVoronoiContainer = this.getCursorVoronoiContainer();
    this.setState({ cursorVoronoiContainer, series });
  };

  private getAdjustedContainerHeight = () => {
    const {
      adjustContainerHeight,
      height,
      containerHeight = height,
      showForecast,
      showInfrastructureLabel,
      showSupplementaryLabel,
    } = this.props;
    const { width } = this.state;

    let adjustedContainerHeight = containerHeight;
    if (adjustContainerHeight) {
      if (showForecast) {
        const maxWidth = showSupplementaryLabel || showInfrastructureLabel ? 900 : 700;
        if (width < maxWidth) {
          adjustedContainerHeight += 25;
        }
      }
    }
    return adjustedContainerHeight;
  };

  private getChart = (series: ChartSeries, index: number) => {
    const { hiddenSeries } = this.state;

    return (
      <ChartArea
        data={!hiddenSeries.has(index) ? series.data : [{ y: null }]}
        interpolation="monotoneX"
        key={series.childName}
        name={series.childName}
        style={series.style}
      />
    );
  };

  // Returns CursorVoronoiContainer component
  private getCursorVoronoiContainer = () => {
    const { formatter, formatOptions } = this.props;

    // Note: Container order is important
    const CursorVoronoiContainer: any = createContainer('voronoi', 'cursor');

    return (
      <CursorVoronoiContainer
        cursorDimension="x"
        labels={({ datum }) => getTooltipLabel(datum, formatter, formatOptions)}
        mouseFollowTooltips
        voronoiDimension="x"
        voronoiPadding={{
          bottom: 50,
          left: 8,
          right: 8,
          top: 8,
        }}
      />
    );
  };

  private getEndDate() {
    const { currentData, forecastData, forecastConeData, previousData } = this.props;
    const previousDate = previousData ? getDate(getDateRange(previousData, true, true)[1]) : 0;
    const currentDate = currentData ? getDate(getDateRange(currentData, true, true)[1]) : 0;
    const forecastDate = forecastData ? getDate(getDateRange(forecastData, true, true)[1]) : 0;
    const forecastConeDate = forecastConeData ? getDate(getDateRange(forecastConeData, true, true)[1]) : 0;

    return currentDate > 0 || previousDate > 0
      ? Math.max(currentDate, forecastDate, forecastConeDate, previousDate)
      : 31;
  }

  // Returns onMouseOver, onMouseOut, and onClick events for the interactive legend
  private getEvents() {
    const { hiddenSeries, series } = this.state;

    const result = getInteractiveLegendEvents({
      chartNames: getChartNames(series),
      isHidden: index => isSeriesHidden(hiddenSeries, index),
      legendName: 'legend',
      onLegendClick: props => this.handleLegendClick(props.index),
    });
    return result;
  }

  private getLegend = () => {
    const { legendItemsPerRow } = this.props;
    const { hiddenSeries, series, width } = this.state;

    // Todo: use PF legendAllowWrap feature
    return (
      <ChartLegend
        data={getLegendData(series, hiddenSeries)}
        gutter={20}
        height={25}
        itemsPerRow={legendItemsPerRow}
        name="legend"
        orientation={width > 150 ? 'horizontal' : 'vertical'}
      />
    );
  };

  // Hide each data series individually
  private handleLegendClick = (index: number) => {
    const hiddenSeries = initHiddenSeries(this.state.series, this.state.hiddenSeries, index);
    this.setState({ hiddenSeries });
  };

  private handleResize = () => {
    const { width } = this.state;
    const { clientWidth = 0 } = this.containerRef.current || {};

    if (clientWidth !== width) {
      this.setState({ width: clientWidth });
    }
  };

  public render() {
    const {
      height,
      intl,
      padding = {
        bottom: 50,
        left: 8,
        right: 8,
        top: 8,
      },
      title,
    } = this.props;
    const { cursorVoronoiContainer, hiddenSeries, series, width } = this.state;
    const domain = getDomain(series, hiddenSeries);
    const endDate = this.getEndDate();
    const midDate = Math.floor(endDate / 2);

    // Clone original container. See https://issues.redhat.com/browse/COST-762
    const container = cursorVoronoiContainer
      ? React.cloneElement(cursorVoronoiContainer, {
          disable: !isDataAvailable(series, hiddenSeries),
          labelComponent: (
            <ChartLegendTooltip
              legendData={getLegendData(series, hiddenSeries, true)}
              title={datum => intl.formatMessage(messages.ChartDayOfTheMonth, { day: datum.x })}
            />
          ),
        })
      : undefined;

    return (
      <>
        <Title headingLevel="h3" size="md">
          {title}
        </Title>
        <div className="chartOverride" ref={this.containerRef} style={{ height: this.getAdjustedContainerHeight() }}>
          <div style={{ height, width }} data-testid="trend-chart-wrapper">
            <Chart
              containerComponent={container}
              domain={domain}
              events={this.getEvents()}
              height={height}
              legendAllowWrap
              legendComponent={this.getLegend()}
              legendData={getLegendData(series, hiddenSeries)}
              legendPosition="bottom-left"
              padding={padding}
              theme={ChartTheme}
              width={width}
            >
              {series &&
                series.map((s, index) => {
                  return this.getChart(s, index);
                })}
              <ChartAxis style={chartStyles.xAxis} tickValues={[1, midDate, endDate]} />
              <ChartAxis dependentAxis style={chartStyles.yAxis} />
            </Chart>
          </div>
        </div>
      </>
    );
  }
}

const TrendChart = injectIntl(TrendChartBase);

export { TrendChart, TrendChartProps };
