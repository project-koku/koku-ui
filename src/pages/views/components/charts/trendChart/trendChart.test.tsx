import { render, screen } from '@testing-library/react';
import { AwsReport, AwsReportData } from 'api/reports/awsReports';
import * as utils from 'pages/views/components/charts/common/chartDatumUtils';
import React from 'react';

import { TrendChart, TrendChartProps } from './trendChart';

const currentMonthReport: AwsReport = createReport('2018-01-15');
const previousMonthReport: AwsReport = createReport('2017-12-15');

const currentData = utils.transformReport(currentMonthReport, utils.ChartType.daily);
const previousData = utils.transformReport(previousMonthReport, utils.ChartType.daily);

const props: TrendChartProps = {
  title: 'Trend Title',
  height: 100,
  intl: null,
  currentData,
  previousData,
  formatter: jest.fn(),
  formatOptions: {},
};

test('reports are formatted to datums', () => {
  render(<TrendChart {...props} />);
  /* eslint-disable-next-line testing-library/no-node-access */
  expect(screen.getByText(props.title).parentElement).toMatchSnapshot();
  expect(screen.queryByText(/no data/i)).toBeNull();
});

test('null previous and current reports are handled', () => {
  render(<TrendChart {...props} currentData={null} previousData={null} />);
  expect(screen.getAllByText(/no data/i).length).toBe(2);
});

test('height from props is used', () => {
  render(<TrendChart {...props} />);
  expect(screen.getByTestId('trend-chart-wrapper').getAttribute('style')).toContain('height: 100px');
});

function createReport(date: string): AwsReport {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, cost = 1): AwsReportData {
  const item = { value: cost, units: 'unit' };
  return {
    date,
    values: [{ date, cost: item }],
  };
}
