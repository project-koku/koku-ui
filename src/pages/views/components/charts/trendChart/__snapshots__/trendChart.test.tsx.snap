// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports are formatted to datums 1`] = `
<div>
  <h3
    class="pf-c-title pf-m-md"
    data-ouia-component-id="OUIA-Generated-Title-1"
    data-ouia-component-type="PF4/Title"
    data-ouia-safe="true"
  >
    Trend Title
  </h3>
  <div
    class="chartOverride"
    style="height: 100px;"
  >
    <div
      data-testid="trend-chart-wrapper"
      style="height: 100px; width: 0px;"
    >
      <div
        class="pf-c-chart"
        style="height: 100%; width: 100%; user-select: none; pointer-events: none; position: relative;"
      >
        <svg
          height="100"
          role="img"
          style="pointer-events: all; width: 100%; height: 100%;"
          viewBox="0 0 0 100"
          width="0"
        >
          <g
            clip-path="url(#victory-clip-5)"
            style="height: 100%; width: 100%; user-select: none;"
          >
            <defs>
              <clippath
                id="victory-clip-5"
              >
                <rect
                  height="42"
                  vector-effect="non-scaling-stroke"
                  width="16"
                  x="-8"
                  y="8"
                />
              </clippath>
            </defs>
            <g>
              <path
                d="M0.5333333333333332,50L0.5333333333333332,50Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: none; stroke: none; fill-opacity: 0.3; stroke-width: 2;"
              />
              <path
                d="M0.5333333333333332,50Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: none; stroke: #d2d2d2; fill-opacity: 0.3; stroke-width: 2;"
              />
            </g>
          </g>
          <g
            clip-path="url(#victory-clip-6)"
            style="height: 100%; width: 100%; user-select: none;"
          >
            <defs>
              <clippath
                id="victory-clip-6"
              >
                <rect
                  height="42"
                  vector-effect="non-scaling-stroke"
                  width="16"
                  x="-8"
                  y="8"
                />
              </clippath>
            </defs>
            <g>
              <path
                d="M0.5333333333333332,50L0.5333333333333332,50Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: none; stroke: none; fill-opacity: 0.3; stroke-width: 2;"
              />
              <path
                d="M0.5333333333333332,50Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: none; stroke: #38812f; fill-opacity: 0.3; stroke-width: 2;"
              />
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: var(--pf-chart-axis--axis--stroke--Color, #d2d2d2); fill: var(--pf-chart-axis--axis--Fill, transparent); stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="8"
              x2="-8"
              y1="50"
              y2="50"
            />
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-2-tickLabels-0"
                x="8"
                y="76.97"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="middle"
                  x="8"
                >
                  1
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-2-tickLabels-1"
                x="0.5333333333333332"
                y="76.97"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="middle"
                  x="0.5333333333333332"
                >
                  15
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-2-tickLabels-2"
                x="-8"
                y="76.97"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="middle"
                  x="-8"
                >
                  31
                </tspan>
              </text>
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: var(--pf-chart-axis--axis--stroke--Color, #d2d2d2); fill: var(--pf-chart-axis--axis--Fill, transparent); stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="8"
              x2="8"
              y1="8"
              y2="50"
            />
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-3-tickLabels-0"
                x="-7"
                y="39.5"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-size: 0px; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="end"
                  x="-7"
                >
                  0.00000000005
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-3-tickLabels-1"
                x="-7"
                y="29"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-size: 0px; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="end"
                  x="-7"
                >
                  0.00000000010
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-3-tickLabels-2"
                x="-7"
                y="18.5"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-size: 0px; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="end"
                  x="-7"
                >
                  0.00000000015
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <text
                direction="inherit"
                dx="0"
                id="chart-axis-3-tickLabels-3"
                x="-7"
                y="8"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="font-size: 0px; font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-axis--tick-label--Fill, #4f5255);"
                  text-anchor="end"
                  x="-7"
                >
                  0.00000000020
                </tspan>
              </text>
            </g>
          </g>
          <g>
            <rect
              height="61.809999999999995"
              role="presentation"
              shape-rendering="auto"
              style="fill: none;"
              vector-effect="non-scaling-stroke"
              width="7392.509375000034"
              x="0"
              y="82"
            />
            <path
              d="M 7.84, 98.156
      h12.32
      v-4.312
      h-12.32
      z"
              role="presentation"
              shape-rendering="auto"
              style="fill: #d2d2d2;"
            />
            <path
              d="M 7.84, 129.061
      h12.32
      v-4.312
      h-12.32
      z"
              role="presentation"
              shape-rendering="auto"
              style="fill: #38812f;"
            />
            <text
              direction="inherit"
              dx="0"
              id="legend-labels-0"
              x="30.8"
              y="100.97"
            >
              <tspan
                dx="0"
                dy="0"
                style="font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-global--label--Fill, #151515);"
                text-anchor="start"
                x="30.8"
              >
                {month, select, 0 {{count, plural, one {Cost (Jan {startDate})} other {Cost (Jan {startDate}-{endDate})}}} 1 {{count, plural, one {Cost (Feb {startDate})} other {Cost (Feb {startDate}-{endDate})}}} 2 {{count, plural, one {Cost (Mar {startDate})} other {Cost (Mar {startDate}-{endDate})}}} 3 {{count, plural, one {Cost (Apr {startDate})} other {Cost (Apr {startDate}-{endDate})}}} 4 {{count, plural, one {Cost (May {startDate})} other {Cost (May {startDate}-{endDate})}}} 5 {{count, plural, one {Cost (Jun {startDate})} other {Cost (Jun {startDate}-{endDate})}}} 6 {{count, plural, one {Cost (Jul {startDate})} other {Cost (Jul {startDate}-{endDate})}}} 7 {{count, plural, one {Cost (Aug {startDate})} other {Cost (Aug {startDate}-{endDate})}}} 8 {{count, plural, one {Cost (Sep {startDate})} other {Cost (Sep {startDate}-{endDate})}}} 9 {{count, plural, one {Cost (Oct {startDate})} other {Cost (Oct {startDate}-{endDate})}}} 10 {{count, plural, one {Cost (Nov {startDate})} other {Cost (Nov {startDate}-{endDate})}}} 11 {{count, plural, one {Cost (Dec {startDate})} other {Cost (Dec {startDate}-{endDate})}}} other {}}{"count":31,"startDate":"1","endDate":"31","month":11,"year":2017}
              </tspan>
            </text>
            <text
              direction="inherit"
              dx="0"
              id="legend-labels-1"
              x="30.8"
              y="131.875"
            >
              <tspan
                dx="0"
                dy="0"
                style="font-family: \\"RedHatText\\",\\"Overpass\\",overpass,helvetica,arial,sans-serif; font-size: 14px; letter-spacing: normal; padding: 10px; stroke: var(--pf-chart-global--label--stroke, transparent); fill: var(--pf-chart-global--label--Fill, #151515);"
                text-anchor="start"
                x="30.8"
              >
                {month, select, 0 {{count, plural, one {Cost (Jan {startDate})} other {Cost (Jan {startDate}-{endDate})}}} 1 {{count, plural, one {Cost (Feb {startDate})} other {Cost (Feb {startDate}-{endDate})}}} 2 {{count, plural, one {Cost (Mar {startDate})} other {Cost (Mar {startDate}-{endDate})}}} 3 {{count, plural, one {Cost (Apr {startDate})} other {Cost (Apr {startDate}-{endDate})}}} 4 {{count, plural, one {Cost (May {startDate})} other {Cost (May {startDate}-{endDate})}}} 5 {{count, plural, one {Cost (Jun {startDate})} other {Cost (Jun {startDate}-{endDate})}}} 6 {{count, plural, one {Cost (Jul {startDate})} other {Cost (Jul {startDate}-{endDate})}}} 7 {{count, plural, one {Cost (Aug {startDate})} other {Cost (Aug {startDate}-{endDate})}}} 8 {{count, plural, one {Cost (Sep {startDate})} other {Cost (Sep {startDate}-{endDate})}}} 9 {{count, plural, one {Cost (Oct {startDate})} other {Cost (Oct {startDate}-{endDate})}}} 10 {{count, plural, one {Cost (Nov {startDate})} other {Cost (Nov {startDate}-{endDate})}}} 11 {{count, plural, one {Cost (Dec {startDate})} other {Cost (Dec {startDate}-{endDate})}}} other {}}{"count":31,"startDate":"1","endDate":"31","month":0,"year":2018}
              </tspan>
            </text>
          </g>
        </svg>
        <div
          style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
        >
          <svg
            height="100"
            style="overflow: visible; width: 100%; height: 100%;"
            viewBox="0 0 0 100"
            width="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
